Userland/Libraries/LibJS/Module.h

```diff
@@ -22,15 +64,34 @@ class Module : public RefCounted<Module> {
     StringView filename() const { return m_filename; }
 
     Environment* environment() { return m_environment.cell(); }
-    Object* namespace_() { return m_namespace.cell(); }
+
+    ThrowCompletionOr<Object*> get_module_namespace(VM& vm);
+
+    virtual ThrowCompletionOr<void> link(VM& vm) = 0;
+    virtual ThrowCompletionOr<Promise*> evaluate(VM& vm) = 0;
+
+    virtual ThrowCompletionOr<Vector<FlyString>> get_exported_names(VM& vm, Vector<Module*> export_star_set = {}) = 0;
+    virtual ThrowCompletionOr<ResolvedBinding> resolve_export(VM& vm, FlyString const& export_name, Vector<ResolvedBinding> resolve_set = {}) = 0;
+
+    virtual ThrowCompletionOr<u32> inner_module_linking(VM& vm, Vector<Module*>& stack, u32 index);
+    virtual ThrowCompletionOr<u32> inner_module_evaluation(VM& vm, Vector<Module*>& stack, u32 index);
 
 protected:
     explicit Module(Realm&, StringView filename);
 
+    void set_environment(Environment* environment)
+    {
+        m_environment = make_handle(environment);
+    }
+
 private:
-    // Handles are not safe unless we keep the VM alive.
-    NonnullRefPtr<VM> m_vm;
```

Comment:
This `RefPtr<VM>` was causing cycles since we (have to) store Modules in the VM also by RefPTR.
Instead we simply take a VM parameter if you want to do anything useful with a Module.
This might become a problem later with the browser but I think we can leave that for another time. 
Also we can always store a reference in [[CustomData]] so it should be extensible enough to work.